alloc.h

```diff
@@ -19,52 +22,124 @@ void* kalloc(size_t sz);
 //    Allocate and return a pointer to at least `sz` contiguous bytes of
 //    memory. Returns `nullptr` if `sz == 0` or on failure.
 //
 //    The caller should initialize the returned memory before using it.
 //    The handout allocator sets returned memory to 0xCC 
+
 void* kalloc(size_t sz) 
 {
-    // REWRITE 
-    if (sz > PAGESIZE) return nullptr;
+    //special case 
+    if (sz == 0) return nullptr;
 
     void* ptr = nullptr;
+    cout << "ALLOCATING" << endl;
 
-    if (next_free_pa < (1UL << 21))
-    {
-        ptr = pa2kptr(next_free_pa);
-        next_free_pa += PAGESIZE;
+    int order = msb(sz - 1);
+    if (order < 12){
+        order = 12; 
+    }
+    int currentOrder = order;
+
+    while (currentOrder <= 21){
+        //empty means FULL!!
+        //special case for 21
+        if (currentOrder == 21 && !orders[21].empty()){
+            cout << "SPECIAL CASE 21" << endl;
+            page* page = orders[currentOrder].pop_front();
+            int index = page->address / 4096;
+            pageArray[index].free = 0;
+            return page;
+        }
+
+        //searching for nearest unoccupied order
+        while (orders[currentOrder].empty()){
+            currentOrder++;
+            if (currentOrder > 21){
+                cout << "no free blocks" << endl;
+                return ptr;
+            }
+        }
+        while (currentOrder >= order){
+            //allocating data
+            if (currentOrder == order){
+                page* p = orders[currentOrder].pop_front();
+                int indexOfP = p->address / 4096;
+                pageArray[indexOfP].free = 0;
+                ptr = p;
+                break;
+            } else { //break blocks if currentOrder > order 
+                page* pa = orders[currentOrder].pop_front();
+                currentOrder--;
+                int indexOfPa = pa->address / 4096;
+                int buddyIndex = getBuddy(indexOfPa * 4096, currentOrder);
+                orders[currentOrder].push_back(&pageArray[buddyIndex]);
+                orders[currentOrder].push_back(&pageArray[indexOfPa]);
+            }
+        }
     }
 
-    if (ptr) memset(ptr, 0xCC, PAGESIZE);
-
+    cout << "ALLOCATED";
     return ptr;
 }
 
-
 // kfree(ptr)
 //    Free a pointer previously returned by `kalloc`. Does nothing if
 //    `ptr == nullptr`.
+//doesnt work THERES A SEGMENTATION FAULT??????
 void kfree(void* ptr) {
     // YOUR CODE HERE
-    cout << "kfree not implemented yet!" << endl;
+    //cout << "kfree not implemented yet!" << endl;
+    cout << "FREEING" << endl;
+    int index = kptr2pa(ptr) / 4096;
+    int order = pageArray[index].order;
+    int buddyIndex = getBuddy(index, order) / 4096;
+    if (order == 21){
+        orders[order].push_back(&pageArray[index]);
+        pageArray[index].free = 1;
+    }
+    if (ptr != nullptr){
+        //if both the page and its buddy are free
+        while (pageArray[index].free == 1 && pageArray[buddyIndex].free == 1){
```

Comment:
shouldn't you be updating the buddyIndex in this loop? Otherwise I believe it will only run once -2pts